File: CIncrement/20/code.cpp
class Solution {
public:
    bool match(char fromStr, char fromStack) {
        if (fromStr == '}') {
            if (fromStack == '{') {
                return true;
            }
        } else if (fromStr == ')') {
            if (fromStack == '(') {
                return true;
            }
        } else if (fromStr == ']') {
            if (fromStack == '[') {
                return true;
            }
        }
        return false;
    }
    bool isOpening(char ch) {
        if (ch == '{' || ch == '(' || ch == '[') {
            return true;
        }
        return false;
    }
    bool isValid(string s) {
        if (s.length() % 2 != 0) {
            return false;
        }
        stack<char> pStack;
        for (string::size_type i = 0; i < s.length(); i++) {
            if (isOpening(s[i])) {
                pStack.push(s[i]);
            } else {
                if (pStack.empty()) {
                    return false;
                }
                if (match(s[i], pStack.top())) {
                    pStack.pop();
                } else {
                    return false;
                }
            }
        }
        return pStack.empty() ? true : false;
    }
};

File: Python3/20/code.py
class Solution:
    def match(self, opening: str, closing: str) -> bool:
        if (opening == "("):
            if (closing == ")"):
                return True
            else: 
                return False
        
        elif (opening == "{"):
            if (closing == "}"):
                return True
            else:
                return False
        
        elif (opening == "["):
            if (closing == "]"):
                return True
            else:
                return False
        
    
    def isOpening(self, s: str) -> bool:
        if (s == "{" or s == "(" or s == "["):
            return True
        else:
            return False
    
    
    def isValid(self, s: str) -> bool:
        if (len(s) % 2 != 0):
            return False
        
        pstack = []
        
        for p in s:
            if (self.isOpening(p) == True):
                pstack.append(p)
                continue
                
            if (len(pstack) == 0):
                # Return false since the stack cannot be empty if a closing bracket is read
                return False 
            else:
                if (self.match(pstack.pop(), p) == False):
                    return False
        
        if (len(pstack) != 0):
            return False
        return True
